commit 37f6e18353e8149d604709df5365b60cde36ae63
Author: Matthew Barnes <mbarnes@redhat.com>
Date:   Thu Jan 20 13:01:49 2011 -0500

    evolution-data-server-2.28.3-crash-in-name-selector-entry.patch

diff --git a/libedataserverui/e-name-selector-entry.c b/libedataserverui/e-name-selector-entry.c
index 0a0c131..ed74097 100644
--- a/libedataserverui/e-name-selector-entry.c
+++ b/libedataserverui/e-name-selector-entry.c
@@ -1222,9 +1222,10 @@ user_insert_text (ENameSelectorEntry *name_selector_entry, gchar *new_text,
 		gtk_editable_insert_text (
 			GTK_EDITABLE (name_selector_entry),
 			new_text, new_text_length, position);
-		post_insert_update (name_selector_entry, *position);
 
 		chars_inserted = *position - old_position;
+		if (chars_inserted > 0)
+			post_insert_update (name_selector_entry, *position);
 
 	/* Otherwise, apply some rules as to where spaces and commas
 	 * can be inserted, and insert a trailing space after comma. */
