diff -up evolution-data-server-3.22.7/addressbook/libedata-book/evolution-addressbook-factory-subprocess.c.gtype-init-workaround evolution-data-server-3.22.7/addressbook/libedata-book/evolution-addressbook-factory-subprocess.c
--- evolution-data-server-3.22.7/addressbook/libedata-book/evolution-addressbook-factory-subprocess.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/addressbook/libedata-book/evolution-addressbook-factory-subprocess.c	2017-11-14 11:40:44.056226796 +0100
@@ -168,6 +168,8 @@ main (gint argc,
 
 	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 #if defined (ENABLE_MAINTAINER_MODE) && defined (HAVE_GTK)
 	if (g_getenv ("EDS_TESTING") == NULL)
diff -up evolution-data-server-3.22.7/calendar/libedata-cal/evolution-calendar-factory-subprocess.c.gtype-init-workaround evolution-data-server-3.22.7/calendar/libedata-cal/evolution-calendar-factory-subprocess.c
--- evolution-data-server-3.22.7/calendar/libedata-cal/evolution-calendar-factory-subprocess.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/calendar/libedata-cal/evolution-calendar-factory-subprocess.c	2017-11-14 11:40:44.056226796 +0100
@@ -165,6 +165,8 @@ main (gint argc,
 
 	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 #if defined (ENABLE_MAINTAINER_MODE) && defined (HAVE_GTK)
 	if (g_getenv ("EDS_TESTING") == NULL)
diff -up evolution-data-server-3.22.7/libedataserver/e-source-registry.c.gtype-init-workaround evolution-data-server-3.22.7/libedataserver/e-source-registry.c
--- evolution-data-server-3.22.7/libedataserver/e-source-registry.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/libedataserver/e-source-registry.c	2017-11-14 11:40:44.056226796 +0100
@@ -1758,6 +1758,8 @@ e_source_registry_new_sync (GCancellable
 	 *     until GObject's type initialization deadlock issue is fixed.
 	 *     Apparently only the synchronous instantiation is affected. */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 	registry = source_registry_dup_uninitialized_singleton ();
 
diff -up evolution-data-server-3.22.7/services/evolution-addressbook-factory/evolution-addressbook-factory.c.gtype-init-workaround evolution-data-server-3.22.7/services/evolution-addressbook-factory/evolution-addressbook-factory.c
--- evolution-data-server-3.22.7/services/evolution-addressbook-factory/evolution-addressbook-factory.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/services/evolution-addressbook-factory/evolution-addressbook-factory.c	2017-11-14 11:40:44.056226796 +0100
@@ -58,6 +58,8 @@ main (gint argc,
 
 	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 #if defined (ENABLE_MAINTAINER_MODE) && defined (HAVE_GTK)
 	if (g_getenv ("EDS_TESTING") == NULL)
diff -up evolution-data-server-3.22.7/services/evolution-calendar-factory/evolution-calendar-factory.c.gtype-init-workaround evolution-data-server-3.22.7/services/evolution-calendar-factory/evolution-calendar-factory.c
--- evolution-data-server-3.22.7/services/evolution-calendar-factory/evolution-calendar-factory.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/services/evolution-calendar-factory/evolution-calendar-factory.c	2017-11-14 11:40:44.056226796 +0100
@@ -60,6 +60,8 @@ main (gint argc,
 
 	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 #if defined (ENABLE_MAINTAINER_MODE) && defined (HAVE_GTK)
 	if (g_getenv ("EDS_TESTING") == NULL)
diff -up evolution-data-server-3.22.7/services/evolution-source-registry/evolution-source-registry.c.gtype-init-workaround evolution-data-server-3.22.7/services/evolution-source-registry/evolution-source-registry.c
--- evolution-data-server-3.22.7/services/evolution-source-registry/evolution-source-registry.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/services/evolution-source-registry/evolution-source-registry.c	2017-11-14 11:40:44.056226796 +0100
@@ -158,6 +158,8 @@ main (gint argc,
 
 	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 #if defined (ENABLE_MAINTAINER_MODE) && defined (HAVE_GTK)
 	if (g_getenv ("EDS_TESTING") == NULL)
diff -up evolution-data-server-3.22.7/services/evolution-user-prompter/evolution-user-prompter.c.gtype-init-workaround evolution-data-server-3.22.7/services/evolution-user-prompter/evolution-user-prompter.c
--- evolution-data-server-3.22.7/services/evolution-user-prompter/evolution-user-prompter.c.gtype-init-workaround	2017-03-20 10:07:46.000000000 +0100
+++ evolution-data-server-3.22.7/services/evolution-user-prompter/evolution-user-prompter.c	2017-11-14 11:40:44.056226796 +0100
@@ -52,6 +52,8 @@ main (gint argc,
 
 	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
 	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
 
 	prompt_user_init (&argc, &argv);
 
