commit 7ae9bda838e954eeaad7d9eafed0bf66d5532838
Author: Matthew Barnes <mbarnes@redhat.com>
Date:   Wed Oct 31 19:39:55 2012 -0400

    libedataserverui-docs.sgml: Remove invalid reference.

diff --git a/docs/reference/libedataserverui/libedataserverui-docs.sgml b/docs/reference/libedataserverui/libedataserverui-docs.sgml
index a86b059..a2f3f58 100644
--- a/docs/reference/libedataserverui/libedataserverui-docs.sgml
+++ b/docs/reference/libedataserverui/libedataserverui-docs.sgml
@@ -15,7 +15,6 @@
     <xi:include href="xml/e-category-completion.xml"/>
     <xi:include href="xml/e-cell-renderer-color.xml"/>
     <xi:include href="xml/e-contact-store.xml"/>
-    <xi:include href="xml/e-data-server-ui-marshal.xml"/>
     <xi:include href="xml/e-destination-store.xml"/>
     <xi:include href="xml/e-name-selector.xml"/>
     <xi:include href="xml/e-name-selector-dialog.xml"/>
