diff -up evolution-data-server-3.12.11/modules/google-backend/module-google-backend.c.goa-google-calendar-auth-method evolution-data-server-3.12.11/modules/google-backend/module-google-backend.c
--- evolution-data-server-3.12.11/modules/google-backend/module-google-backend.c.goa-google-calendar-auth-method	2016-04-13 13:51:48.862364981 +0200
+++ evolution-data-server-3.12.11/modules/google-backend/module-google-backend.c	2016-04-13 13:51:48.956364978 +0200
@@ -104,13 +104,16 @@ G_DEFINE_DYNAMIC_TYPE (
 	E_TYPE_COLLECTION_BACKEND_FACTORY)
 
 static void
-google_backend_calendar_update_auth_method (ESource *source)
+google_backend_calendar_update_auth_method (ESource *child_source,
+					    ESource *master_source)
 {
 	EOAuth2Support *oauth2_support;
 	ESourceAuthentication *auth_extension;
 	const gchar *method;
 
-	oauth2_support = e_server_side_source_ref_oauth2_support (E_SERVER_SIDE_SOURCE (source));
+	oauth2_support = e_server_side_source_ref_oauth2_support (E_SERVER_SIDE_SOURCE (child_source));
+	if (!oauth2_support && master_source)
+		oauth2_support = e_server_side_source_ref_oauth2_support (E_SERVER_SIDE_SOURCE (master_source));
 
 	/* The host name and WebDAV resource path depend on the
 	 * authentication method used, so update those here too. */
@@ -121,25 +124,35 @@ google_backend_calendar_update_auth_meth
 		method = "plain/password";
 	}
 
-	auth_extension = e_source_get_extension (source, E_SOURCE_EXTENSION_AUTHENTICATION);
+	auth_extension = e_source_get_extension (child_source, E_SOURCE_EXTENSION_AUTHENTICATION);
 	e_source_authentication_set_method (auth_extension, method);
 
 	g_clear_object (&oauth2_support);
 }
 
 static void
-google_backend_contacts_update_auth_method (ESource *source)
+google_backend_calendar_update_auth_method_cb (ESource *child_source,
+					       GParamSpec *param,
+					       EBackend *backend)
+{
+	google_backend_calendar_update_auth_method (child_source, e_backend_get_source (backend));
+}
+
+static void
+google_backend_contacts_update_auth_method (ESource *child_source,
+					    ESource *master_source)
 {
 	EOAuth2Support *oauth2_support;
 	ESourceAuthentication *extension;
 	const gchar *extension_name;
 	const gchar *method;
 
-	oauth2_support = e_server_side_source_ref_oauth2_support (
-		E_SERVER_SIDE_SOURCE (source));
+	oauth2_support = e_server_side_source_ref_oauth2_support (E_SERVER_SIDE_SOURCE (child_source));
+	if (!oauth2_support && master_source)
+		oauth2_support = e_server_side_source_ref_oauth2_support (E_SERVER_SIDE_SOURCE (master_source));
 
 	extension_name = E_SOURCE_EXTENSION_AUTHENTICATION;
-	extension = e_source_get_extension (source, extension_name);
+	extension = e_source_get_extension (child_source, extension_name);
 	method = (oauth2_support != NULL) ? "OAuth2" : "ClientLogin";
 	e_source_authentication_set_method (extension, method);
 
@@ -147,6 +160,14 @@ google_backend_contacts_update_auth_meth
 }
 
 static void
+google_backend_contacts_update_auth_method_cb (ESource *child_source,
+					       GParamSpec *param,
+					       EBackend *backend)
+{
+	google_backend_contacts_update_auth_method (child_source, e_backend_get_source (backend));
+}
+
+static void
 google_add_uid_to_hashtable (gpointer source,
 			     gpointer known_sources)
 {
@@ -267,6 +288,8 @@ google_add_found_source (ECollectionBack
 			child_webdav = e_source_get_extension (source, E_SOURCE_EXTENSION_WEBDAV_BACKEND);
 			resource = e_source_get_extension (source, E_SOURCE_EXTENSION_RESOURCE);
 
+			google_backend_calendar_update_auth_method (source, master_source);
+
 			e_source_authentication_set_user (child_auth, e_source_collection_get_identity (collection_extension));
 			e_source_webdav_set_soup_uri (child_webdav, uri);
 			e_source_resource_set_identity (resource, identity);
@@ -463,6 +486,8 @@ google_populate_thread (gpointer data)
 	else
 		calendar_url = "https://www.google.com/calendar/dav/";
 
+	google_backend_calendar_update_auth_method (source, NULL);
+
 	if (e_source_collection_get_calendar_enabled (collection_extension) && calendar_url &&
 	    e_webdav_discover_sources_sync (collection, calendar_url,
 		&discovered_sources, NULL, NULL, &local_error)) {
@@ -669,11 +694,11 @@ google_backend_child_added (ECollectionB
 	 *     Many-to-one property bindinds tend not to work so well. */
 	extension_name = E_SOURCE_EXTENSION_CALENDAR;
 	if (e_source_has_extension (child_source, extension_name)) {
-		google_backend_calendar_update_auth_method (child_source);
+		google_backend_calendar_update_auth_method (child_source, collection_source);
 		g_signal_connect (
 			child_source, "notify::oauth2-support",
-			G_CALLBACK (google_backend_calendar_update_auth_method),
-			NULL);
+			G_CALLBACK (google_backend_calendar_update_auth_method_cb),
+			backend);
 	}
 
 	/* Keep the contacts authentication method up-to-date.
@@ -683,11 +708,11 @@ google_backend_child_added (ECollectionB
 	 *     Many-to-one property bindings tend not to work so well. */
 	extension_name = E_SOURCE_EXTENSION_ADDRESS_BOOK;
 	if (e_source_has_extension (child_source, extension_name)) {
-		google_backend_contacts_update_auth_method (child_source);
+		google_backend_contacts_update_auth_method (child_source, collection_source);
 		g_signal_connect (
 			child_source, "notify::oauth2-support",
-			G_CALLBACK (google_backend_contacts_update_auth_method),
-			NULL);
+			G_CALLBACK (google_backend_contacts_update_auth_method_cb),
+			backend);
 	}
 }
 
