diff -up evolution-data-server-3.12.11/Makefile.am.truly-disable-gtk-doc evolution-data-server-3.12.11/Makefile.am
--- evolution-data-server-3.12.11/Makefile.am.truly-disable-gtk-doc	2014-07-10 12:27:18.000000000 +0200
+++ evolution-data-server-3.12.11/Makefile.am	2015-05-28 19:29:13.077569079 +0200
@@ -20,11 +20,14 @@ SUBDIRS = \
 	modules \
 	services \
 	tests \
-	docs \
 	art \
 	po \
 	$(NULL)
 
+if ENABLE_GTK_DOC
+SUBDIRS += docs
+endif
+
 if BUILD_EXAMPLES
 SUBDIRS += examples
 endif
