diff -up evolution-data-server-2.32.3/addressbook/libebook/e-book.c.cal-factory-close-on-logout evolution-data-server-2.32.3/addressbook/libebook/e-book.c
--- evolution-data-server-2.32.3/addressbook/libebook/e-book.c.cal-factory-close-on-logout	2015-01-07 21:45:18.320747720 +0100
+++ evolution-data-server-2.32.3/addressbook/libebook/e-book.c	2015-01-07 21:45:31.680747163 +0100
@@ -333,7 +333,6 @@ e_book_activate (GError **error)
 	}
 
 	connection = g_dbus_proxy_get_connection (G_DBUS_PROXY (book_factory_proxy));
-	g_dbus_connection_set_exit_on_close (connection, FALSE);
 	book_connection_closed_id = g_dbus_connection_signal_subscribe (connection,
 		NULL,						/* sender */
 		"org.freedesktop.DBus",				/* interface */
diff -up evolution-data-server-2.32.3/addressbook/libedata-book/e-data-book-factory.c.cal-factory-close-on-logout evolution-data-server-2.32.3/addressbook/libedata-book/e-data-book-factory.c
--- evolution-data-server-2.32.3/addressbook/libedata-book/e-data-book-factory.c.cal-factory-close-on-logout	2015-01-07 21:44:12.533750462 +0100
+++ evolution-data-server-2.32.3/addressbook/libedata-book/e-data-book-factory.c	2015-01-07 21:44:12.664750457 +0100
@@ -572,6 +572,11 @@ on_name_lost (GDBusConnection *connectio
 	}
 
 	g_mutex_unlock (factory->priv->connections_lock);
+
+	if (g_strcmp0 (name, E_DATA_BOOK_FACTORY_SERVICE_NAME) == 0) {
+		g_print ("Quitting on the D-Bus name lost...\n");
+		g_main_loop_quit (loop);
+	}
 }
 
 #ifndef G_OS_WIN32
diff -up evolution-data-server-2.32.3/calendar/libecal/e-cal.c.cal-factory-close-on-logout evolution-data-server-2.32.3/calendar/libecal/e-cal.c
--- evolution-data-server-2.32.3/calendar/libecal/e-cal.c.cal-factory-close-on-logout	2015-01-07 21:44:39.033749358 +0100
+++ evolution-data-server-2.32.3/calendar/libecal/e-cal.c	2015-01-07 21:44:58.440748549 +0100
@@ -709,7 +709,6 @@ e_cal_activate (GError **error)
 	}
 
 	connection = g_dbus_proxy_get_connection (G_DBUS_PROXY (cal_factory_proxy));
-	g_dbus_connection_set_exit_on_close (connection, FALSE);
 	cal_connection_closed_id = g_dbus_connection_signal_subscribe (connection,
 		NULL,						/* sender */
 		"org.freedesktop.DBus",				/* interface */
diff -up evolution-data-server-2.32.3/calendar/libedata-cal/e-data-cal-factory.c.cal-factory-close-on-logout evolution-data-server-2.32.3/calendar/libedata-cal/e-data-cal-factory.c
--- evolution-data-server-2.32.3/calendar/libedata-cal/e-data-cal-factory.c.cal-factory-close-on-logout	2015-01-07 21:44:12.534750462 +0100
+++ evolution-data-server-2.32.3/calendar/libedata-cal/e-data-cal-factory.c	2015-01-07 21:44:12.664750457 +0100
@@ -876,6 +876,11 @@ on_name_lost (GDBusConnection *connectio
 		g_list_foreach (copy, remove_data_cal_cb, NULL);
 		g_list_free (copy);
 	}
+
+	if (g_strcmp0 (name, E_DATA_CAL_FACTORY_SERVICE_NAME) == 0) {
+		g_print ("Quitting on the D-Bus name lost...\n");
+		g_main_loop_quit (loop);
+	}
 }
 
 #ifndef G_OS_WIN32
