diff -up evolution-data-server-3.22.6/camel/providers/pop3/camel-pop3-folder.c.maybe-uninitialized-variable evolution-data-server-3.22.6/camel/providers/pop3/camel-pop3-folder.c
--- evolution-data-server-3.22.6/camel/providers/pop3/camel-pop3-folder.c.maybe-uninitialized-variable	2017-03-14 15:24:48.131584612 +0100
+++ evolution-data-server-3.22.6/camel/providers/pop3/camel-pop3-folder.c	2017-03-14 15:25:42.541580816 +0100
@@ -409,7 +409,7 @@ pop3_folder_get_message_internal_sync (C
 	CamelPOP3Command *pcr;
 	CamelPOP3FolderInfo *fi;
 	gchar buffer[1];
-	gint i, last;
+	gint i = -1, last;
 	CamelStream *stream = NULL;
 	CamelService *service;
 	CamelSettings *settings;
