diff -up evolution-data-server-2.32.3/addressbook/backends/google/util.c.google-book-other-fax evolution-data-server-2.32.3/addressbook/backends/google/util.c
--- evolution-data-server-2.32.3/addressbook/backends/google/util.c.google-book-other-fax	2013-07-24 07:55:48.775435182 +0200
+++ evolution-data-server-2.32.3/addressbook/backends/google/util.c	2013-07-24 07:55:48.784435182 +0200
@@ -509,7 +509,6 @@ static const struct RelTypeMap rel_type_
 	{ "callback", { EVC_X_CALLBACK, NULL }},
 	{ "car", { "CAR", NULL }},
 	{ "company_main", {EVC_X_COMPANY, NULL }},
-	{ "fax", { "FAX", NULL }},
 	{ "isdn", { "ISDN", NULL }},
 	{ "main", { "PREF", NULL }},
 	{ "mobile", { "CELL", NULL }},
@@ -518,7 +517,12 @@ static const struct RelTypeMap rel_type_
 	{ "pager", { "PAGER", NULL }},
 	{ "radio", { EVC_X_RADIO, NULL }},
 	{ "telex", { EVC_X_TELEX, NULL }},
-	{ "tty_tdd", { EVC_X_TTYTDD, NULL }}
+	{ "tty_tdd", { EVC_X_TTYTDD, NULL }},
+
+	/* XXX This has no clear mapping to an EContact field.
+	 * It's listed here for completeness, but ordered
+	 * last so that "other_fax" is preferred. */
+	{ "fax", { "FAX", NULL }}
 };
 
 static const struct RelTypeMap rel_type_map_im[] = {
