diff -up evolution-data-server-3.12.11/addressbook/libedata-book/e-book-sqlite.c.book-sqlite-left-outer-join evolution-data-server-3.12.11/addressbook/libedata-book/e-book-sqlite.c
--- evolution-data-server-3.12.11/addressbook/libedata-book/e-book-sqlite.c.book-sqlite-left-outer-join	2015-06-19 10:58:15.805862676 +0200
+++ evolution-data-server-3.12.11/addressbook/libedata-book/e-book-sqlite.c	2015-06-19 10:58:15.826862675 +0200
@@ -5347,7 +5347,7 @@ ebsql_generate_select (EBookSqlite *ebsq
 				 *     WHERE email_list.value LIKE "boogieman%"
 				 */
 				ebsql_string_append_printf (
-					string, " JOIN %Q AS %s ON +%s.uid = summary.uid",
+					string, " LEFT OUTER JOIN %Q AS %s ON +%s.uid = summary.uid",
 					field->aux_table,
 					field->aux_table_symbolic,
 					field->aux_table_symbolic);
