diff -up evolution-data-server-2.32.3/camel/providers/imap/camel-imap-store.c.tls-for-imaps evolution-data-server-2.32.3/camel/providers/imap/camel-imap-store.c
--- evolution-data-server-2.32.3/camel/providers/imap/camel-imap-store.c.tls-for-imaps	2011-04-21 21:35:36.000000000 +0200
+++ evolution-data-server-2.32.3/camel/providers/imap/camel-imap-store.c	2013-06-11 17:50:06.745370642 +0200
@@ -415,7 +415,7 @@ enum {
 };
 
 #ifdef CAMEL_HAVE_SSL
-#define SSL_PORT_FLAGS (CAMEL_TCP_STREAM_SSL_ENABLE_SSL2 | CAMEL_TCP_STREAM_SSL_ENABLE_SSL3)
+#define SSL_PORT_FLAGS (CAMEL_TCP_STREAM_SSL_ENABLE_SSL2 | CAMEL_TCP_STREAM_SSL_ENABLE_SSL3 | CAMEL_TCP_STREAM_SSL_ENABLE_TLS)
 #define STARTTLS_FLAGS (CAMEL_TCP_STREAM_SSL_ENABLE_TLS)
 #endif
 
diff -up evolution-data-server-2.32.3/camel/providers/imapx/camel-imapx-server.c.tls-for-imaps evolution-data-server-2.32.3/camel/providers/imapx/camel-imapx-server.c
--- evolution-data-server-2.32.3/camel/providers/imapx/camel-imapx-server.c.tls-for-imaps	2013-06-11 17:50:06.704370643 +0200
+++ evolution-data-server-2.32.3/camel/providers/imapx/camel-imapx-server.c	2013-06-11 17:50:06.747370642 +0200
@@ -357,7 +357,7 @@ enum {
 	USE_SSL_WHEN_POSSIBLE
 };
 
-#define SSL_PORT_FLAGS (CAMEL_TCP_STREAM_SSL_ENABLE_SSL2 | CAMEL_TCP_STREAM_SSL_ENABLE_SSL3)
+#define SSL_PORT_FLAGS (CAMEL_TCP_STREAM_SSL_ENABLE_SSL2 | CAMEL_TCP_STREAM_SSL_ENABLE_SSL3 | CAMEL_TCP_STREAM_SSL_ENABLE_TLS)
 #define STARTTLS_FLAGS (CAMEL_TCP_STREAM_SSL_ENABLE_TLS)
 
 static gboolean imapx_select(CamelIMAPXServer *is, CamelFolder *folder, gboolean force, GError **error);
