diff -up evolution-data-server-2.28.3/calendar/backends/caldav/e-cal-backend-caldav.c.caldav-uri-encoding evolution-data-server-2.28.3/calendar/backends/caldav/e-cal-backend-caldav.c
--- evolution-data-server-2.28.3/calendar/backends/caldav/e-cal-backend-caldav.c.caldav-uri-encoding	2010-03-01 02:03:50.000000000 -0500
+++ evolution-data-server-2.28.3/calendar/backends/caldav/e-cal-backend-caldav.c	2012-09-07 09:29:12.053868535 -0400
@@ -2175,8 +2175,20 @@ initialize_backend (ECalBackendCalDAV *c
 
 		/* properly encode uri */
 		if (suri && suri->path) {
-			gchar *tmp = soup_uri_encode (suri->path, NULL);
-			gchar *path = soup_uri_normalize (tmp, "/");
+			gchar *tmp, *path;
+
+			if (suri->path && strchr (suri->path, '%')) {
+				/* If path contains anything already encoded, then decode it first,
+				   thus it'll be managed properly. For example, the '#' in a path
+				   is in URI shown as %23 and not doing this decode makes it being
+				   like %2523, which is not what is wanted here. */
+				tmp = soup_uri_decode (suri->path);
+				soup_uri_set_path (suri, tmp);
+				g_free (tmp);
+			}
+
+			tmp = soup_uri_encode (suri->path, NULL);
+			path = soup_uri_normalize (tmp, "/");
 
 			soup_uri_set_path (suri, path);
 
