diff -up evolution-data-server-2.32.3/libebackend/e-file-cache.c.file-cache-freeze evolution-data-server-2.32.3/libebackend/e-file-cache.c
--- evolution-data-server-2.32.3/libebackend/e-file-cache.c.file-cache-freeze	2011-04-21 21:35:37.000000000 +0200
+++ evolution-data-server-2.32.3/libebackend/e-file-cache.c	2013-06-11 15:02:14.069372935 +0200
@@ -34,7 +34,7 @@ struct _EFileCachePrivate {
 	gchar *filename;
 	EXmlHash *xml_hash;
 	gboolean dirty;
-	gboolean frozen;
+	guint32 frozen;
 };
 
 /* Property IDs */
@@ -159,7 +159,7 @@ e_file_cache_init (EFileCache *cache)
 
 	priv = g_new0 (EFileCachePrivate, 1);
 	priv->dirty = FALSE;
-	priv->frozen = FALSE;
+	priv->frozen = 0;
 	cache->priv = priv;
 }
 
@@ -463,7 +463,8 @@ e_file_cache_freeze_changes (EFileCache 
 
 	priv = cache->priv;
 
-	priv->frozen = TRUE;
+	priv->frozen++;
+	g_return_if_fail (priv->frozen > 0);
 }
 
 /**
@@ -481,8 +482,10 @@ e_file_cache_thaw_changes (EFileCache *c
 
 	priv = cache->priv;
 
-	priv->frozen = FALSE;
-	if (priv->dirty) {
+	g_return_if_fail (priv->frozen > 0);
+
+	priv->frozen--;
+	if (!priv->frozen && priv->dirty) {
 		e_xmlhash_write (priv->xml_hash);
 		priv->dirty = FALSE;
 	}
